portal/src/components/roles-and-groups/AddRoleGroupsDialog.tsx

```diff
@@ -0,0 +1,232 @@
+    roleGroups,
+  }) {
+    const { renderToString } = useContext(MessageContext);
+    const existingGroupIDs = useMemo(() => {
+      return new Set(roleGroups.map((group) => group.id));
+    }, [roleGroups]);
+
+    const [groupTags, setGroupTags] = useState<GroupTag[]>([]);
+
+    const { addRoleToGroups, loading, error } = useAddRoleToGroupsMutation();
+    const { refetch: refetchRole } = useRoleQuery(roleID, {
+      skip: true,
+    });
+    const { refetch } = useQuery<
+      GroupsListQueryQuery,
+      GroupsListQueryQueryVariables
+    >(GroupsListQueryDocument);
+
+    const onChangeGroupTags = useCallback((tags?: ITag[]) => {
+      if (tags === undefined) {
+        setGroupTags([]);
+      } else {
+        setGroupTags(tags as GroupTag[]);
+      }
+    }, []);
+    const [searchGroupKeyword, setSearchGroupKeyword] = useState<string>("");
+    const onSearchInputChange = useCallback((value: string): string => {
+      setSearchGroupKeyword(value);
+      return value;
+    }, []);
+    const onClearGroupTags = useCallback(() => setGroupTags([]), []);
+
+    const onResolveGroupSuggestions = useCallback(
+      async (filter: string, selectedTags?: ITag[]): Promise<ITag[]> => {
+        const selectedGroupIDs = new Set(
+          selectedTags?.map((tag) => (tag as GroupTag).group.id)
+        );
+        const result = await refetch({ searchKeyword: filter, pageSize: 100 });
+
+        if (result.data.groups?.edges == null) {
+          return [];
+        }
+        return result.data.groups.edges.flatMap<GroupTag>((edge) => {
+          const node = edge?.node;
+          if (node == null) {
+            return [];
+          }
+          // Filter out existing groups
+          if (existingGroupIDs.has(node.id)) {
+            return [];
+          }
+
+          // Filter out selected groups
+          if (selectedGroupIDs.has(node.id)) {
+            return [];
+          }
+
+          return [toGroupTag(node)];
+        });
+      },
+      [existingGroupIDs, refetch]
+    );
+
+    const onDialogDismiss = useCallback(() => {
+      if (loading || isHidden) {
+        return;
+      }
+      onDismiss();
+    }, [isHidden, loading, onDismiss]);
+
+    const onSubmit = useCallback(() => {
+      if (loading || isHidden || groupTags.length === 0) {
+        return;
+      }
+      addRoleToGroups(
+        roleKey,
+        groupTags.map((tag) => tag.group.key)
+      )
+        .then(async () => {
+          // Update the cache
+          return refetchRole({ roleID: roleID });
+        })
+        .then(
+          () => onDismiss(),
+          (e: unknown) => {
+            onDismiss();
+            throw e;
+          }
+        );
+    }, [
+      loading,
+      isHidden,
+      groupTags,
+      addRoleToGroups,
+      roleKey,
+      refetchRole,
+      roleID,
+      onDismiss,
+    ]);
+
+    useEffect(() => {
+      // Reset states on isHidden changes
+      setGroupTags([]);
+    }, [isHidden]);
+
```

Comment:
I think it's more intuitive to just put this in onDismiss?

Reply:
onDimiss doesn't really mean it is dismissed, it is actually a "request of dismiss".
The flag that controls it is `isHidden`.

Reply:
How about `onDimissed`? Instead of resetting tags when the flag is toggle (and also adding a extra useEffect), resetting the tags when the dialog is dimissed seems eaiser to understand.